# Aufgabenblatt 4

## Lineare Optimierung - 4


In [2]:
import numpy as np
import math
import copy

In [135]:
def simplex(a,b,c,zf_wert):
    first=True
    arr_pivot_spalte=[]
    arr_choose_pivot=[]
    arr_division=[]
    arr_pivot_zeile=[]
    
    pivot_help=np.min(c)
    if(pivot_help>=0):
        print('zielfunktion fertig')
        print('Zielfunktionswert: ', round(zf_wert,2))
        print('c: ')
        for i in range(0, c.size):
            print(round(c[i],2))
        print('A: ', a)
        print('b: ')
        for i in range(0, b.size):
            print(round(b[i],2))
        return
        
        
    else:
        for j in range(0, c.size):
            if(c[j]==pivot_help):
                pivot_spalte=j
                for i in range(0, b.size):
                    arr_pivot_spalte.append(a[i,j])
        
        for i in range(0, b.size):
            if(arr_pivot_spalte[i] > 0):
                arr_division.append(b[i]/arr_pivot_spalte[i])
                
            arr_choose_pivot.append(arr_pivot_spalte[i])
                
        if(len(arr_division)==0):
            print('no positive values in column: ', pivot_spalte)
            print('Zielfunktionswert: ', round(zf_wert,2))
            print('c: ')
            for i in range(0, c.size):
                print(round(c[i],2))
            print('A: ', a)
            print('b: ')
            for i in range(0, b.size):
                print(round(b[i],2))
            return
        
        #print('pivot spalte elems: ', arr_choose_pivot)
        #choosing pivot elem
        #choosing pivot line
        pivot_zeile=np.min(arr_division)
        for i in range(0, b.size):
            if(arr_pivot_spalte[i]>0):
                if(b[i]/arr_pivot_spalte[i]==pivot_zeile):
                    pivot=arr_choose_pivot[i]
                    pivot_zeile=i
                    break
                
        #adding values to pivot line array
        for i in range(0, b.size):
            if(i==pivot_zeile):
                for j in range(0, c.size):
                    arr_pivot_zeile.append(a[i,j])
                
        #changing b in not pivot lines
        c_pivot_spalte_help=copy.copy(c[pivot_spalte])
        for i in range(0, b.size):
            if(i!=pivot_zeile):
                b[i]=b[i]-arr_pivot_spalte[i]/pivot*b[pivot_zeile]
                #print(b[i], i)
                
                if(first==True):
                    #print('c values: ')
                    for j in range(0, c.size):
                        help=arr_pivot_zeile[j]/pivot*c_pivot_spalte_help
                        #print('help: ', help, 'c: ', c[j])
                        c[j]=c[j]-help
                        #print(c[j], arr_pivot_zeile[j], pivot, c_pivot_spalte_help)
                    first=False
                
        #print('pivot spalte nr: ', pivot_spalte)
        #print('pivot zeile nr: ', pivot_zeile)
        #changing A values in pivot line and zf_wert + b value in line of pivot
        for i in range(0, b.size):
            if(i==pivot_zeile):
                for j in range(0, c.size):
                    a[i,j]=a[i,j]/pivot
                    
                    if(j==pivot_spalte):
                        zf_wert=zf_wert-b[i]/pivot*c_pivot_spalte_help
                    
                b[i]=b[i]/pivot
        
        #print('pivot spalte: ', arr_pivot_spalte)
        #print('pivot zeile: ', arr_pivot_zeile)
        #print('pivot element: ', pivot)
        
        #updating A matrix
        for i in range(0, b.size):
            if(i!=pivot_zeile):
                for j in range(0, c.size):
                    a[i,j]=a[i,j]-arr_pivot_spalte[i]/pivot*arr_pivot_zeile[j]
                        
        #print('check')
        #print('Zielfunktionswert: ', round(zf_wert,2))
        #print('c: ')
        #for i in range(0, c.size):
        #    print(round(c[i],2))
        #print('A: ', a)
        #print('b: ')
        #for i in range(0, b.size):
        #    print(round(b[i],2))
        simplex(a,b,c,zf_wert)
    

### Aufgabe 1

Schreiben Sie für die Aufgaben  a) und b) die Dualprobleme auf und lösen Sie diese mit dem Simplex-Algorithmus.

a) $ x_1 + 2x_2 + x_4 + x_6 \rightarrow \text{max(min) }$

unter den Nebenbedingungen:

$ x_1 + 2x_2 - x_3 + 2x_4 -x_5 + 2x_6 =0,$

$x_1 - 4x_2 + 2 x_3 + 2x_4 - 4 x_6 = 1,$

$ 2x_1 - 2x_2 + x_3 + 4x_4 + x_5 - 2x_6 = 3, $

$ x_i  \geq 0,\,i = 1, \dots, 6 $




Lösung:

Maximierung:
$$
    y_3-y_4+3y_5-3y_6 \rightarrow \text{max} \\
    -y_1+y_2-y_3+y_4-2y_5+2y_6 \leq 1 \\
    -2y_1+2y_2+4y_3-4y_4+2y_5-2y_6 \leq 2 \\
    y_1-y_2-2y_3+3y_4-y_5+y_6 \leq 0 \\
    -2y_1+2y_2-2y_3+2y_4-4y_5+4y_6 \leq 1 \\
    y_1-y_2-y_5+y_6 \leq 0 \\
    -2y_1+2y_2+4y_3-4y_4+2y_5-2y_6 \leq 1
$$

In [136]:
a=np.array([[-1., 1., -1., 1., -2., 2., 1., 0., 0., 0., 0., 0.],
            [-2., 2., 4., -4., 2., -2., 0., 1., 0., 0., 0., 0.],
            [1., -1., -2., 2., -1., 1., 0., 0., 1., 0., 0., 0.],
            [-2., 2., -2., 2., -4., 4., 0., 0., 0., 1., 0., 0.],
            [1., -1., 0., 0., -1., 1., 0., 0., 0., 0., 1., 0.],
            [-2., 2., 4., -4., 2., -2., 0., 0., 0., 0., 0., 1.]])
b=np.array([1., 2., 0., 1., 0., 1.])
c=np.array([0., 0., -1., 1., -3., 3., 0., 0., 0., 0., 0., 0.])
zf_wert=0.
simplex(a,b,c,zf_wert)

no positive values in column:  3
Zielfunktionswert:  1.5
c: 
-3.0
3.0
5.0
-5.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
1.5
A:  [[-3.   3.   3.  -3.   0.   0.   1.   0.   0.   0.   0.   1. ]
 [ 0.   0.   0.   0.   0.   0.   0.   1.   0.   0.   0.  -1. ]
 [ 0.   0.   0.   0.   0.   0.   0.   0.   1.   0.   0.   0.5]
 [-6.   6.   6.  -6.   0.   0.   0.   0.   0.   1.   0.   2. ]
 [ 0.   0.   2.  -2.   0.   0.   0.   0.   0.   0.   1.   0.5]
 [-1.   1.   2.  -2.   1.  -1.   0.   0.   0.   0.   0.   0.5]]
b: 
2.0
1.0
0.5
3.0
0.5
0.5


Antwort: das Problem ist unbeschränkt.

Minimierung:

$$
    -y_3+y_4-3y_5+3y_6 \rightarrow \text{max} \\
    -y_1+y_2-y_3+y_4-2y_5+2y_6 \leq 1 \\
    -2y_1+2y_2+4y_3-4y_4+2y_5-2y_6 \leq 2 \\
    y_1-y_2-2y_3+3y_4-y_5+y_6 \leq 0 \\
    -2y_1+2y_2-2y_3+2y_4-4y_5+4y_6 \leq 1 \\
    y_1-y_2-y_5+y_6 \leq 0 \\
    -2y_1+2y_2+4y_3-4y_4+2y_5-2y_6 \leq 1
$$

In [137]:
a=np.array([[-1., 1., -1., 1., -2., 2., 1., 0., 0., 0., 0., 0.],
            [-2., 2., 4., -4., 2., -2., 0., 1., 0., 0., 0., 0.],
            [1., -1., -2., 2., -1., 1., 0., 0., 1., 0., 0., 0.],
            [-2., 2., -2., 2., -4., 4., 0., 0., 0., 1., 0., 0.],
            [1., -1., 0., 0., -1., 1., 0., 0., 0., 0., 1., 0.],
            [-2., 2., 4., -4., 2., -2., 0., 0., 0., 0., 0., 1.]])
b=np.array([1., 2., 0., 1., 0., 1.])
c=np.array([0., 0., 1., -1., 3., -3., 0., 0., 0., 0., 0., 0.])
zf_wert=0.
simplex(a,b,c,zf_wert)

zielfunktion fertig
Zielfunktionswert:  0.5
c: 
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.5
1.0
0.0
A:  [[ 0.          0.          0.          0.          0.          0.
   1.          0.          0.         -0.5         0.          0.        ]
 [ 0.          0.          0.          0.          0.          0.
   0.          1.          2.          0.          0.          0.        ]
 [ 0.          0.          0.          0.         -1.          1.
   0.          0.         -0.16666667  0.16666667  0.5         0.        ]
 [-1.          1.          0.          0.          0.          0.
   0.          0.         -0.16666667  0.16666667 -0.5         0.        ]
 [ 0.          0.          1.         -1.          0.          0.
   0.          0.         -0.5         0.          0.5         0.        ]
 [ 0.          0.          0.          0.          0.          0.
   0.          0.          2.          0.          0.          1.        ]]
b: 
0.5
2.0
0.17
0.17
0.0
1.0


Antwort: der minimale Zielfunktionswert ist 0.5.


b) $ 3x_1 + 10x_2 + 8x_3 -6 x_4 \rightarrow \text{min(max) }$

unter den Nebenbedingungen:

$ x_1 + 3x_2 + 3 x_3 -  2x_4 = -1,$

$3x_1+ 2x_2 + x_3 - x_4  \geq -1,$

$ x_1 \leq 4,$

$ x_i  \geq 0, \, i = 1, \dots, 4 $

Lösung:

Maximierung:

$$
    -y_1+y_2+y_3+4y_4-Ms_1 \rightarrow \text{max} \\
    -y_1+y_2+3y_3-y_4+y_5 = 3 \\
    -3y_1+3y_2+2y_3+y_6 = 10 \\
    -3y_1+3y_2+3y_3+y_7 = 8 \\
    -2y_1+2y_2+y_3-y_8+s_1 = 6
$$

In [138]:
#M = 10000
a=np.array([[-1., 1., 3., -1., 1., 0., 0., 0., 0.],
            [-3., 3., 2., 0., 0., 1., 0., 0., 0.],
            [-3., 3., 3., 0., 0., 0., 1., 0., 0.],
            [-2., 2., 1., 0., 0., 0., 0., -1., 1.]])
b=np.array([3., 10., 8., 6.])
c=np.array([20001., -20001., -10001., -4., 0., 0., 0., 10000., 0.])
zf_wert=-60000.
simplex(a,b,c,zf_wert)

no positive values in column:  3
Zielfunktionswert:  -6664.0
c: 
0.0
0.0
10000.0
-4.0
0.0
0.0
6667.0
10000.0
0.0
A:  [[ 0.          0.          2.         -1.          1.          0.
  -0.33333333  0.          0.        ]
 [ 0.          0.         -1.          0.          0.          1.
  -1.          0.          0.        ]
 [-1.          1.          1.          0.          0.          0.
   0.33333333  0.          0.        ]
 [ 0.          0.         -1.          0.          0.          0.
  -0.66666667 -1.          1.        ]]
b: 
0.33
2.0
2.67
0.67


Antwort: das Problem ist unbeschränkt.

Minimierung:

$$
    y_1-y_2-y_3-4y_4-Ms_1 \rightarrow \text{max} \\
    -y_1+y_2+3y_3-y_4+y_5 = 3 \\
    -3y_1+3y_2+2y_3+y_6 = 10 \\
    -3y_1+3y_2+3y_3+y_7 = 8 \\
    -2y_1+2y_2+y_3-y_8+s_1 = 6
$$

In [139]:
#M = 10000
a=np.array([[-1., 1., 3., -1., 1., 0., 0., 0., 0.],
            [-3., 3., 2., 0., 0., 1., 0., 0., 0.],
            [-3., 3., 1., 0., 0., 0., 1., 0., 0.],
            [-2., 2., 1., 0., 0., 0., 0., -1., 1.]])
b=np.array([3., 10., 8., 6.])
c=np.array([19999., -19999., -9999., 4., 0., 0., 0., 10000., 0.])
zf_wert=-60000.
simplex(a,b,c,zf_wert)

zielfunktion fertig
Zielfunktionswert:  -24.0
c: 
0.0
0.0
0.0
0.0
4.0
3322.0
3343.0
10000.0
0.0
A:  [[ 0.          0.          1.          0.          0.          1.
  -1.          0.          0.        ]
 [ 0.          0.          0.          1.         -1.          2.66666667
  -2.33333333  0.          0.        ]
 [-1.          1.          0.          0.          0.         -0.33333333
   0.66666667  0.          0.        ]
 [ 0.          0.          0.          0.          0.         -0.33333333
  -0.33333333 -1.          1.        ]]
b: 
2.0
5.0
2.0
0.0


Antwort: der minimale Zielfunktionswert ist -24.

### Aufgabe 2 

Lösen Sie das LP-Problem mit dem Simplex-Algorithmus:

\begin{array}{rcl}
1000 x_1 + 200 x_2 + 400 x_3 &\to& \max\\
x_1 + x_2 + 8 x_3 &\leq& 250\\
20 x_1  + 2 x_2 + x_3&\leq& 200\\
x_1, x_2, x_3 &\geq& 0
\end{array}

a) Um wieviel kann der optimale Zielfunktionswert erhöht werden, wenn man die rechte Seite der ersten Nebenbedingung von 250 auf 251 erhöht?

b) Ermitteln Sie die Schattenpreise und interpretieren Sie diese Werte.

c) Schreiben Sie das duale System auf.  

d) Was ist die Lösung des dualen Systems?

e) Ermitteln Sie die Schattenpreise des dualen Systems.

f) Vergleichen Sie beide Lösungen. Welche Zusammenhänge in Bezug auf Dualität können Sie feststellen?

g) Prüfen Sie Ihre händischen Lösungen von a) - f) mit dem Python-Programm linprog

In [140]:
a=np.array([[1., 1., 8., 1., 0.],
            [20., 2., 1., 0., 1.]])
b=np.array([250., 200.])
c=np.array([-1000., -200., -400., 0., 0.])
zf_wert=0.
simplex(a,b,c,zf_wert)

zielfunktion fertig
Zielfunktionswert:  26000.0
c: 
640.0
0.0
0.0
40.0
80.0
A:  [[-1.2         0.          1.          0.13333333 -0.06666667]
 [10.6         1.          0.         -0.06666667  0.53333333]]
b: 
20.0
90.0


Der maximale Zielfunktionswert ist 26000. Der optimale Punkt ist (0,90,20,0,0).

a) Der Zielfunktionswert wird um 40 erhöht.

b) Schattenpreise: 40, 80. Wenn die erste NB von 250 auf 251 erhöht wird, wird der Zielfunktionswert um 40 größer. Wenn die zweite NB von 200 auf 201 erhöht wird, wird der Zielfunktionswert um 80 größer.

c)

$$
  250y_1+200y_2 \rightarrow \text{min} \\
  y_1+20y_2 \geq 1000 \\
  y_1+2y_2 \geq 200 \\
  8y_1+y_2 \geq 400
$$

$$
  -250y_1-200y_2-Ms_1-Ms_2-Ms_3 \rightarrow \text{max} \\
  y_1+20y_2-y_3+s_1 = 1000 \\
  y_1+2y_2-y_4+s_2 = 200 \\
  8y_1+y_2-y_5+s_3 = 400
$$

In [141]:
#M = 10000
a=np.array([[1., 20., -1., 0., 0., 1., 0., 0.],
            [1., 2., 0., -1., 0., 0., 1., 0.],
            [8., 1., 0., 0., -1., 0., 0., 1]])
b=np.array([1000., 200., 400])
c=np.array([-99750., -229800., 10000., 10000., 10000., 0., 0., 0.])
zf_wert=-16000000.
simplex(a,b,c,zf_wert)

zielfunktion fertig
Zielfunktionswert:  -26000.0
c: 
0.0
0.0
0.0
90.0
20.0
10000.0
9910.0
9980.0
A:  [[  0.           1.           0.          -0.53333333   0.06666667
    0.           0.53333333  -0.06666667]
 [  0.           0.           1.         -10.6          1.2
   -1.          10.6         -1.2       ]
 [  1.           0.           0.           0.06666667  -0.13333333
    0.          -0.06666667   0.13333333]]
b: 
80.0
640.0
40.0


d) Der minimale Wert der Zielfunktion ist 26000. Der optimale Punkt ist (40, 80, 640, 0, 0).

e) Schattenpreise: 0, 90, 20. Wenn die erste NB von 1000 auf 1001 erhöht wird, wird der Zielfunktionswert um 0 größer. Wenn die zweite NB von 200 auf 201 erhöht wird, wird der Zielfunktionswert um 90 größer. Wenn die dritte NB von 400 auf 401 erhöht wird, wird der Zielfunktionswert um 20 größer.

### Aufgabe 3
	
Gegeben ist das folgende Transportproblem (Produktionsstätten $Q_1$, $Q_2$; Verbraucherstätten $S_1$,  $S_2$; [ 2 x 2 ]- Kostenmatrix):

In [2]:
import pandas as pd
import numpy as np
row_labels=['Q1','Q2', 'Nachfrage']
df = pd.DataFrame(data = np.array([[30,20,10],[40,25,20],[15,15,0]]), columns = ['S1','S2','Angebot'], index=row_labels)
df

S1  S2  Angebot
Q1         30  20       10
Q2         40  25       20
Nachfrage  15  15        0

Formulieren Sie das Problem als LP und lösen Sie es durch Dualisierung.

Lösung:

$$
    30x_1+20x_2+40x_3+25x_4 \rightarrow \text{min} \\
    x_1+x_2 = 10 \\
    x_3+x_4 = 20 \\
    x_1+x_3 = 15 \\
    x_2+x_4 = 15
$$

$$
    10y_1-10y_2+20y_3-20y_4+15y_5-15y_6+15y_7-15y_8 \rightarrow \text{max} \\
    y_1-y_2+y_5-y_6 \leq 30 \\
    y_1-y_2+y_7-y_8 \leq 20 \\
    y_3-y_4+y_5-y_6 \leq 40 \\
    y_3-y_4+y_7-y_8 \leq 25
$$

In [142]:
a=np.array([[1., -1., 0., 0., 1., -1., 0., 0., 1., 0., 0., 0.],
            [1., -1., 0., 0., 0., 0., 1., -1., 0., 1., 0., 0.],
            [0., 0., 1., -1., 1., -1., 0., 0., 0., 0., 1., 0.],
            [0., 0., 1., -1., 0., 0., 1., -1., 0., 0., 0., 1.]])
b=np.array([30., 20., 40., 25.])
c=np.array([-10., 10., -20., 20., -15., 15., -15., 15., 0., 0., 0., 0.])
zf_wert=0.
simplex(a,b,c,zf_wert)

zielfunktion fertig
Zielfunktionswert:  875.0
c: 
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
10.0
0.0
5.0
15.0
A:  [[ 1. -1.  0.  0.  0.  0.  1. -1.  1.  0. -1.  1.]
 [ 0.  0.  0.  0.  0.  0.  0.  0. -1.  1.  1. -1.]
 [ 0.  0.  0.  0.  1. -1. -1.  1.  0.  0.  1. -1.]
 [ 0.  0.  1. -1.  0.  0.  1. -1.  0.  0.  0.  1.]]
b: 
15.0
5.0
15.0
25.0


Antwort: Der minimale Zielfunktionswert für das primale Problem ist 875. Der optimale Punkt ist (15,0,25,0,15,0,0,0,0,5,0,0).

### Aufgabe 4
Lösen Sie das LP-Problem


\begin{array}{rcl}
5 x_5 + 2x_4   &\to& \max\\
2x_1 - 4x_2 + 2 x_3 + x_4 - 7x_5  &  = & 18\\
x_1   -  x_2  -2 x_3 - 5 x_4 + 6 x_5 & = & -19\\
2x_2   - 4 x_3  -7 x_4 + 13 x_5 & = & -4\\
x_1, x_2, x_3, x_4,x_5 &\geq& 0
\end{array} 

   

Lösung:

$$
    2x_4+5x_5-Ms_1-Ms_2-Ms_3 \rightarrow \text{max} \\
    2x_1-4x_2+2x_3+x_4-7x_5+x_6 = 18 \\
    2x_1-4x_2+2x_3+x_4-7x_5-x_7+s_1 = 18 \\
    -x_1+x_2+2x_3+5x_4-6x_5-x_8+s_2 = 19 \\
    -x_1+x_2+2x_3+5x_4-6x_5+x_9 = 19 \\
    -2x_2+4x_3+7x_4-13x_5-x_{10}+s_3 = 4 \\
    -2x_2+4x_3+7x_4-13x_5+x_{11} = 4 \\
$$

In [143]:
#M = 10000
a=np.array([[2., -4., 2., 1., -7., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
            [2., -4., 2., 1., -7., 0., -1., 0., 0., 0., 0., 1., 0., 0.],
            [-1., 1., 2., 5., -6., 0., 0., -1., 0., 0., 0., 0., 1., 0.],
            [-1., 1., 2., 5., -6., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
            [0., -2., 4., 7., -13., 0., 0., 0., 0., -1., 0., 0., 0., 1.],
            [0., -2., 4., 7., -13., 0., 0., 0., 0., 0., 1., 0., 0., 0.]])
b=np.array([18., 18., 19., 19., 4., 4.])
c=np.array([-10000., 50000., -80000., -130002., 259995., 0., 10000., 10000., 0., 10000., 0., 0., 0., 0.])
zf_wert=-410000.
simplex(a,b,c,zf_wert)

zielfunktion fertig
Zielfunktionswert:  500.0
c: 
0.0
10.0
0.0
3.0
0.0
10010.0
10000.0
0.0
20.0
15.0
0.0
0.0
10000.0
9985.0
A:  [[ 1.  -1.   0.  -1.   0.   1.   0.   0.   1.   1.   0.   0.   0.  -1. ]
 [ 0.   0.   0.   0.   0.  -1.  -1.   0.   0.   0.   0.   1.   0.   0. ]
 [ 0.   2.   0.   1.   1.   2.   0.   0.   4.   3.   0.   0.   0.  -3. ]
 [ 0.   0.   0.   0.   0.   0.   0.   1.   1.   0.   0.   0.  -1.   0. ]
 [ 0.   6.   1.   5.   0.   6.5  0.   0.  13.   9.5  0.   0.   0.  -9.5]
 [ 0.   0.   0.   0.   0.   0.   0.   0.   0.   1.   1.   0.   0.  -1. ]]
b: 
33.0
0.0
100.0
0.0
326.0
0.0


Antwort: Der maximale Zielfunktionswert ist 500. Der optimale Punkt ist (33, 0, 326, 0, 100, 0, 0, 0, 0, 0, 0).

### Aufgabe 5

Lösen Sie das LP-Problem

\begin{array}{rcl}
-2 x_1 - x_2 - 7 x_3   &\to& \min\\
x_1  + 2x_2 + 3 x_3   &  \leq & 4\\
- x_1   - 4 x_2 + 10 x_3   &  \leq & 7\\
x_1, x_2, x_3 &\geq& 0
\end{array} 

Lösung:

$$
    2x_1+x_2+7x_3 \rightarrow \text{max} \\
    x_1+2x_2+3x_3+x_4 = 4 \\
    -x_1-4x_2+10x_3+x_5 = 7
$$

In [144]:
a=np.array([[1., 2., 3., 1., 0.],
            [-1., -4., 10., 0., 1.]])
b=np.array([4., 7.])
c=np.array([-2., -1., -7., 0., 0.])
zf_wert=0.
simplex(a,b,c,zf_wert)

zielfunktion fertig
Zielfunktionswert:  8.85
c: 
0.0
2.85
0.0
2.08
0.08
A:  [[ 1.          2.46153846  0.          0.76923077 -0.23076923]
 [ 0.         -0.15384615  1.          0.07692308  0.07692308]]
b: 
1.46
0.85


Antwort: der minimale Zielfunktionswert ist -8.85. Der optimale Punkt ist (1.46, 0, 0.85, 0, 0).